src/services/refactors/convertStringOrTemplateLiteral.ts

```diff
@@ -0,0 +1,190 @@
+        if (isParenthesizedExpression(node.parent) && isBinaryExpression(node.parent.parent)) return node.parent.parent;
+        return node;
+    }
+
+    function getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
+        const { file, startPosition } = context;
+        const node = getNodeOrParentOfParentheses(file, startPosition);
+
+        switch (actionName) {
+            case toTemplateLiteralActionName:
+                return { edits: getEditsForToTemplateLiteral(context, node) };
+
+            case toStringConcatenationActionName:
+                return { edits: getEditsForToStringConcatenation(context, node) };
+
+            default:
+                return Debug.fail("invalid action");
+        }
+    }
+
+    function getEditsForToTemplateLiteral(context: RefactorContext, node: Node) {
+        const maybeBinary = getParentBinaryExpression(node);
+        const arrayOfNodes = transformTreeToArray(maybeBinary);
+        const templateLiteral = nodesToTemplate(arrayOfNodes);
+        return textChanges.ChangeTracker.with(context, t => t.replaceNode(context.file, maybeBinary, templateLiteral));
+    }
+
+    function getEditsForToStringConcatenation(context: RefactorContext, node: Node) {
+        const templateLiteral = findAncestor(node, n => isTemplateLiteral(n))! as TemplateLiteral;
+
+        if (isTemplateExpression(templateLiteral)) {
+            const { head, templateSpans } = templateLiteral;
+            const arrayOfNodes = templateSpans.map(templateSpanToExpressions)
+                                              .reduce((accumulator, nextArray) => accumulator.concat(nextArray));
+
+            if (head.text.length !== 0) arrayOfNodes.unshift(createStringLiteral(head.text));
+
+            const binaryExpression = arrayToTree(arrayOfNodes);
+            return textChanges.ChangeTracker.with(context, t => t.replaceNode(context.file, templateLiteral, binaryExpression));
+        }
+        else {
+            const stringLiteral = createStringLiteral(templateLiteral.text);
+            return textChanges.ChangeTracker.with(context, t => t.replaceNode(context.file, node, stringLiteral));
+        }
+    }
+
+    function templateSpanToExpressions(templateSpan: TemplateSpan): Expression[] {
+        const { expression, literal } = templateSpan;
+        const text = literal.text;
+        return text.length === 0 ? [expression] : [expression, createStringLiteral(text)];
+    }
+
+    function isNotEqualsOperator(node: BinaryExpression) {
+        return node.operatorToken.kind !== SyntaxKind.EqualsToken;
+    }
+
+    function getParentBinaryExpression(expr: Node) {
+        while (isBinaryExpression(expr.parent) && isNotEqualsOperator(expr.parent)) {
+            expr = expr.parent;
+        }
+        return expr;
+    }
+
+    function arrayToTree(nodes: ReadonlyArray<Expression>, accumulator?: BinaryExpression): BinaryExpression {

```

Comment:
i think this function would be better as a for loop or taking an index as parameter to avoid the unnecessary calls to `slice`.